File: src/risper/retranscribe.py
from __future__ import annotations

import argparse
import sys
from pathlib import Path

from .clipboard import copy_text
from .config import load_config
from .models import active_profile
from .paste import attempt_paste
from .sessions import append_event, find_session, missing_audio_message, update_metadata
from .sounds import play
from .transcriber import transcribe
from .util import append_log, notify


def retranscribe_session(session_id: str, copy: bool = False, paste: bool = False) -> int:
    config = load_config()
    metadata = find_session(config, session_id)
    if not metadata:
        print(f"No such session: {session_id}", file=sys.stderr)
        return 1

    audio_path = Path(str(metadata["audio_path"]))
    if not audio_path.exists():
        print(missing_audio_message(metadata), file=sys.stderr)
        return 1

    session_dir = audio_path.parent
    status_log = session_dir / "status.log"
    error_log = session_dir / "error.log"
    profile = active_profile(config)
    metadata = update_metadata(
        metadata,
        status="transcribing",
        transcription_engine=profile.engine,
        model=profile.model,
        language=profile.language,
    )
    append_log(status_log, "starting retranscription")
    append_event(
        metadata,
        "retranscription.starting",
        profile=profile.id,
        engine=profile.engine,
        model=profile.model,
        language=profile.language,
    )
    play(config, "transcription_start")

    try:
        transcript = transcribe(
            config,
            audio_path,
            Path(str(metadata["transcript_raw_path"])),
            Path(str(metadata["transcript_clean_path"])),
            profile=profile,
        )
    except Exception as exc:
        message = f"retranscription failed: {exc}"
        append_log(error_log, message)
        append_log(status_log, message)
        append_event(metadata, "retranscription.failed", error=str(exc), error_type=exc.__class__.__name__)
        errors = list(metadata.get("errors", []))
        errors.append(message)
        update_metadata(metadata, status="failed", errors=errors)
        notify("⚠ Risper retranscription failed", "Audio was kept; see session error log.")
        play(config, "error")
        return 1

    append_event(
        metadata,
        "retranscription.completed",
        raw_path=str(metadata["transcript_raw_path"]),
        clean_path=str(metadata["transcript_clean_path"]),
        transcript_chars=len(transcript),
    )
    paste_attempted = False
    paste_succeeded = False
    errors = [err for err in list(metadata.get("errors", [])) if "transcription failed:" not in str(err)]

    if copy or paste:
        copied, clipboard_message = copy_text(transcript)
        append_log(status_log, clipboard_message)
        append_event(metadata, "clipboard.copy", ok=copied, message=clipboard_message, transcript_chars=len(transcript))
        if not copied:
            errors.append(clipboard_message)
            update_metadata(metadata, status="failed", errors=errors)
            play(config, "error")
            return 1
        if paste:
            paste_attempted = True
            append_event(
                metadata,
                "paste.attempting",
                mode=config.paste_mode,
                session_type=metadata.get("session_type"),
            )
            paste_succeeded, paste_message = attempt_paste(config)
            append_log(status_log, paste_message)
            confirmation = (
                "helper_returned_success_target_unverified"
                if paste_succeeded
                else "not_pasted_clipboard_retained"
            )
            append_event(
                metadata,
                "paste.result",
                ok=paste_succeeded,
                mode=config.paste_mode,
                session_type=metadata.get("session_type"),
                message=paste_message,
                confirmation=confirmation,
            )
            if not paste_succeeded:
                errors.append(paste_message)
        else:
            confirmation = "not_attempted"
    else:
        confirmation = "not_attempted"

    status = "complete" if not paste_attempted else "paste_attempted" if paste_succeeded else "paste_failed"
    update_metadata(
        metadata,
        status=status,
        paste_attempted=paste_attempted,
        paste_helper_succeeded=paste_succeeded,
        paste_succeeded=False if paste_attempted else paste_succeeded,
        paste_confirmation=confirmation,
        errors=errors,
    )
    play(config, "error" if status == "paste_failed" else "success")
    print(transcript)
    return 0


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Retranscribe a saved Risper session.")
    parser.add_argument("session_id", nargs="?", default="last", help="Session id, or 'last'.")
    parser.add_argument("--copy", action="store_true", help="Copy transcript after retranscribing.")
    parser.add_argument("--paste", action="store_true", help="Copy and attempt paste after retranscribing.")
    args = parser.parse_args(argv)
    return retranscribe_session(args.session_id, copy=args.copy or args.paste, paste=args.paste)


if __name__ == "__main__":
    raise SystemExit(main())

File: retranscribe/retranscribe.go
// Package retranscribe re-runs a saved session through the active local
// model. The Go workflow is clipboard-only: there is no paste helper path.
package retranscribe

import (
	"errors"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/rob137/risper/config"
	"github.com/rob137/risper/desktop"
	"github.com/rob137/risper/events"
	"github.com/rob137/risper/internal/args"
	"github.com/rob137/risper/models"
	"github.com/rob137/risper/session"
	"github.com/rob137/risper/spend"
	"github.com/rob137/risper/transcription"
)

func Main(argv []string) int {
	parser := flag.NewFlagSet("risper retranscribe", flag.ContinueOnError)
	parser.SetOutput(os.Stderr)
	if err := parser.Parse(args.Reorder(argv, nil)); err != nil {
		if err == flag.ErrHelp {
			return 0
		}
		return 2
	}
	if parser.NArg() > 1 {
		fmt.Fprintln(os.Stderr, "risper retranscribe: unexpected positional arguments")
		return 2
	}
	sessionID := "last"
	if parser.NArg() == 1 {
		sessionID = parser.Arg(0)
	}
	cfg, err := config.Load()
	if err != nil {
		return reportError(err)
	}
	metadata, err := session.Find(cfg, sessionID)
	if err != nil {
		return reportError(err)
	}
	if metadata == nil {
		fmt.Fprintln(os.Stderr, "No such session: "+sessionID)
		return 1
	}
	audioPath := selectedAudioPath(metadata)
	if _, err := os.Stat(audioPath); err != nil {
		fmt.Fprintln(os.Stderr, session.MissingAudioMessage(metadata))
		return 1
	}
	profile, err := models.Active(cfg)
	if err != nil {
		return failSession(cfg, metadata, err)
	}
	profiles, err := models.Load(cfg)
	if err != nil {
		return failSession(cfg, metadata, err)
	}
	fallback, err := models.SelectFallback(profile, profiles, cfg.VoiceTriggerProfile)
	if err != nil {
		return failSession(cfg, metadata, err)
	}
	return run(cfg, metadata, profile, fallback, audioPath)
}

func selectedAudioPath(metadata *session.Metadata) string {
	if metadata == nil {
		return ""
	}
	// The mixed file is the durable transcription contract for both current
	// and older sessions. Per-source tracks remain available for future tools.
	return metadata.AudioPath
}

func run(cfg config.Config, metadata *session.Metadata, profile, fallback models.Profile, audioPath string) int {
	dir := session.SessionDir(metadata)
	metadata.Status = "transcribing"
	metadata.TranscriptionEngine = profile.Engine
	metadata.Model = profile.Model
	metadata.Language = profile.Language
	if err := session.SaveMetadata(metadata); err != nil {
		return reportError(err)
	}
	appendLog(filepath.Join(dir, session.StatusLogFile), "starting retranscription audio_source=mixed")
	_, _ = events.Append(dir, "retranscription.starting", map[string]any{
		"profile": profile.ID, "engine": profile.Engine, "model": profile.Model,
		"language": profile.Language, "audio_path": audioPath, "audio_source": "mixed",
	})
	desktop.Notify(cfg, "📝 Retranscribing speech", "Using "+profile.ID+".")
	desktop.Play(cfg, "transcription_start")
	if err := transcription.Start(cfg, metadata, profile.ID); err != nil {
		return failSession(cfg, metadata, err)
	}
	defer transcription.Finish(cfg)
	onProcessStart := func(pid int) error {
		if err := transcription.SetWorkerPID(cfg, pid); err != nil {
			return err
		}
		if fallback.ID != "" {
			desktop.Notify(cfg, "📝 Retranscribing speech", "OpenAI unavailable; using "+fallback.ID+" locally.")
		}
		return nil
	}
	result := transcription.TranscriptionResult{Profile: profile}
	var err error
	if fallback.ID != "" {
		result, err = transcription.TranscribeWithFallback(
			profile, fallback, audioPath, metadata.TranscriptRawPath, metadata.TranscriptCleanPath,
			time.Duration(profile.FallbackTimeoutSeconds)*time.Second, onProcessStart,
		)
	} else {
		result.Transcript, err = transcription.Transcribe(
			profile, audioPath, metadata.TranscriptRawPath, metadata.TranscriptCleanPath, onProcessStart,
		)
	}
	if err != nil {
		return failSession(cfg, metadata, fmt.Errorf("retranscription failed: %w", err))
	}
	transcript := result.Transcript
	actualProfile := result.Profile
	metadata.TranscriptionEngine = actualProfile.Engine
	metadata.Model = actualProfile.Model
	metadata.Language = actualProfile.Language
	spend.RecordEstimate(metadata, actualProfile.Engine, actualProfile.BillingPricePerMinute, actualProfile.BillingCurrency)
	if result.PrimaryError != nil {
		fallbackMessage := fmt.Sprintf("OpenAI profile %s unavailable; used local profile %s: %v", profile.ID, actualProfile.ID, result.PrimaryError)
		_ = appendLog(filepath.Join(dir, session.StatusLogFile), fallbackMessage)
		_, _ = events.Append(dir, "retranscription.fallback", map[string]any{
			"from_profile": profile.ID, "from_engine": profile.Engine,
			"to_profile": actualProfile.ID, "to_engine": actualProfile.Engine,
			"reason": result.PrimaryError.Error(),
		})
	}
	_, _ = events.Append(dir, "retranscription.completed", map[string]any{
		"raw_path": metadata.TranscriptRawPath, "clean_path": metadata.TranscriptCleanPath,
		"transcript_chars": len([]rune(transcript)), "audio_source": "mixed",
		"profile": actualProfile.ID, "engine": actualProfile.Engine,
		"model": actualProfile.Model, "requested_profile": profile.ID,
		"used_fallback": result.PrimaryError != nil,
	})

	copied, clipboardMessage := desktop.CopyText(transcript)
	appendLog(filepath.Join(dir, session.StatusLogFile), clipboardMessage)
	_, _ = events.Append(dir, "clipboard.copy", map[string]any{
		"ok": copied, "message": clipboardMessage, "transcript_chars": len([]rune(transcript)),
	})
	if !copied {
		return failSession(cfg, metadata, errors.New(clipboardMessage))
	}
	appendLog(filepath.Join(dir, session.StatusLogFile), "automatic paste skipped; transcript left on clipboard")
	_, _ = events.Append(dir, "paste.skipped", map[string]any{
		"reason": "automatic_paste_disabled", "session_type": metadata.SessionType,
	})
	falseValue := false
	metadata.Status = "complete"
	metadata.Errors = []string{}
	metadata.PasteAttempted = &falseValue
	metadata.PasteHelperSucceeded = &falseValue
	metadata.PasteSucceeded = &falseValue
	metadata.PasteConfirmation = "not_attempted_automatic_paste_disabled"
	if err := session.SaveMetadata(metadata); err != nil {
		return reportError(err)
	}
	desktop.Notify(cfg, "✅ Risper copied", "Transcript is on the clipboard.")
	desktop.Play(cfg, "success")
	fmt.Println(transcript)
	return 0
}

func failSession(cfg config.Config, metadata *session.Metadata, err error) int {
	message := err.Error()
	if !strings.HasPrefix(message, "retranscription failed:") {
		message = "retranscription failed: " + message
	}
	dir := session.SessionDir(metadata)
	appendLog(filepath.Join(dir, session.ErrorLogFile), message)
	appendLog(filepath.Join(dir, session.StatusLogFile), message)
	_, _ = events.Append(dir, "retranscription.failed", map[string]any{
		"error": strings.TrimPrefix(message, "retranscription failed: "), "error_type": fmt.Sprintf("%T", err),
	})
	metadata.Status = "failed"
	metadata.Errors = append(metadata.Errors, message)
	if saveErr := session.SaveMetadata(metadata); saveErr != nil {
		fmt.Fprintln(os.Stderr, "risper retranscribe:", saveErr)
		return 1
	}
	desktop.Notify(cfg, "⚠ Risper retranscription failed", "Audio was kept; see session error log.")
	desktop.Play(cfg, "error")
	fmt.Fprintln(os.Stderr, message)
	return 1
}

func appendLog(path, message string) error {
	handle, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0o644)
	if err != nil {
		return err
	}
	defer handle.Close()
	_, err = fmt.Fprintf(handle, "%s %s\n", nowRFC3339(), message)
	return err
}

func nowRFC3339() string {
	return timeNow().Format("2006-01-02T15:04:05Z07:00")
}

var timeNow = func() time.Time { return time.Now() }

func reportError(err error) int {
	fmt.Fprintln(os.Stderr, "risper retranscribe:", err)
	return 1
}

// Codex gpt-5.6-sol, xhigh, prompted by Robert Kirby
